hazelcast/src/main/java/com/hazelcast/aggregation/Aggregator.java

```diff
@@ -0,0 +1,85 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.aggregation;
+
+import java.io.Serializable;
+import java.util.Map;
+
+/**
+ * Base class for all aggregators. Exposes API for parallel two-phase aggregations:
+ * - accumulation of entries by multiple instance of aggregators
+ * - combining all aggregators into one to calculate the final result
+ * <p>
+ * Aggregator does not have to be thread-safe.
+ * accumulate() and combine() calls may be interwoven.
+ * <p>
+ * The very instance passed to an aggregate() method will not be used at all. It is just a prototype object
+ * that will be cloned using serialization, since each partition gets its own instance of an aggregator.
+ * In this way the aggregator is not used by multiple-threads. Each thread gets its own aggregator instance.
+ *
+ * @param <R> aggregation result
+ * @param <K> entry key type
+ * @param <V> entry value type
+ */
+public abstract class Aggregator<R, K, V> implements Serializable {
+
+    /**
+     * Accumulates the given entries.
+     *
+     * @param entry entries to accumulate.
+     */
+    public abstract void accumulate(Map.Entry<K, V> entry);
+
+    /**
+     * Called after the last call to combine on a specific instance. Enables disposing of the intermediary state.
+     * This should be a very fast operation that just disposes unnecessary state (if applicable).
+     * <p>
+     * IMPORTANT: It may not be called if the instance aggregator does not take part in the accumulation phase.
+     * It's caused by the fact that the aggregation may be run in a parallel way and each thread gets a clone of the
+     * aggregator.
+     */
+    public void onAccumulationFinished() {
+    }
+
+    /**
+     * Incorporates the intermediary result of the given aggregator to this instance of the aggregator.
+     * Enables merging the intermediary state from a given aggregator.
+     * It is used when the aggregation is split into a couple of aggregators.
+     *
+     * @param aggregator aggregator providing intermediary results to be combined into the results of this aggregator.
+     */
+    public abstract void combine(Aggregator aggregator);
+
+    /**
+     * Called after the last call to combine on a specific instance. Enables disposing of the intermediary state.
+     * This should be a very fast operation that just disposes unnecessary state (if applicable).
+     * <p>
+     * IMPORTANT: It may not be called if the instance aggregator does not take part in the combination phase.
+     * It's caused by the fact that the aggregation may be run in a parallel way and each thread gets a clone of the
+     * aggregator.
+     */
+    public void onCombinationFinished() {
```

Comment:
@jerrinot I see your point, didn't consider this. In this case I wound't change it. 👍